In [2]:
import pandas

In [14]:
genomes = pandas.read_parquet('genomes.parquet')
print(f"There are {len(genomes)} genomes in the dataset.")
genomes[:3]

There are 77 genomes in the dataset.


species  n_lineages   lineage  \
uniqueid                                                                        
site.ENA.subj.SAMN03648077.lab.1.iso.1  M. tuberculosis           1  lineage4   
site.ENA.subj.SAMEA1403592.lab.1.iso.1  M. tuberculosis           1  lineage4   
site.ENA.subj.SAMN07766210.lab.1.iso.1  M. tuberculosis           1  lineage1   

                                            sublineage  mycobacterial_reads  \
uniqueid                                                                      
site.ENA.subj.SAMN03648077.lab.1.iso.1  lineage4.6.1.2              4820884   
site.ENA.subj.SAMEA1403592.lab.1.iso.1    lineage4.2.1              3290666   
site.ENA.subj.SAMN07766210.lab.1.iso.1    lineage1.2.1              3667176   

                                        tb_reads  tb_coverage  tb_depth  \
uniqueid                                                                  
site.ENA.subj.SAMN03648077.lab.1.iso.1   4695252      99.7035  103.2910   
site.ENA.subj.SAMEA1403592.lab.1.iso.1   3213994      99.8773   69.0550   
site.ENA.subj.SAMN07766210.lab.1.iso.1   3585232      99.7362   99.2324   

                                        mykrobe_lineage_coverage  \
uniqueid                                                           
site.ENA.subj.SAMN03648077.lab.1.iso.1                     100.0   
site.ENA.subj.SAMEA1403592.lab.1.iso.1                     100.0   
site.ENA.subj.SAMN07766210.lab.1.iso.1                       0.0   

                                        mykrobe_lineage_depth     antibiogram  \
uniqueid                                                                        
site.ENA.subj.SAMN03648077.lab.1.iso.1                   90.0  RRSSSSUSSSRSUS   
site.ENA.subj.SAMEA1403592.lab.1.iso.1                   64.0  SSSSSSSSSSSSSS   
site.ENA.subj.SAMN07766210.lab.1.iso.1                    0.0  SSSSSSUSSSRSUS   

                                       pipeline_build  
uniqueid                                               
site.ENA.subj.SAMN03648077.lab.1.iso.1        149a518  
site.ENA.subj.SAMEA1403592.lab.1.iso.1        149a518  
site.ENA.subj.SAMN07766210.lab.1.iso.1        149a518

In [15]:
genomes.info()

<class 'pandas.core.frame.DataFrame'>
Index: 77 entries, site.ENA.subj.SAMN03648077.lab.1.iso.1 to site.ENA.subj.SAMEA1016056.lab.1.iso.1
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype   
---  ------                    --------------  -----   
 0   species                   77 non-null     object  
 1   n_lineages                77 non-null     int64   
 2   lineage                   77 non-null     category
 3   sublineage                77 non-null     category
 4   mycobacterial_reads       77 non-null     int64   
 5   tb_reads                  77 non-null     int64   
 6   tb_coverage               77 non-null     float64 
 7   tb_depth                  77 non-null     float64 
 8   mykrobe_lineage_coverage  76 non-null     float64 
 9   mykrobe_lineage_depth     76 non-null     float64 
 10  antibiogram               77 non-null     category
 11  pipeline_build            77 non-null     category
dtypes: category(4), float64(4), int64(3), ob

In [5]:
master = pandas.read_parquet('ENA_LOOKUP.parquet')
master[:3]

study_accession sample_accession  \
UNIQUEID                                                                    
site.05.subj.PTAN-0001.lab.TAN-001.iso.1      PRJEB23245   SAMEA104362043   
site.05.subj.PTAN-0002.lab.TAN-004.iso.1      PRJEB23245   SAMEA104362044   
site.05.subj.PTAN-0003.lab.TAN-006.iso.1      PRJEB23245   SAMEA104362045   

                                         run_accession  \
UNIQUEID                                                 
site.05.subj.PTAN-0001.lab.TAN-001.iso.1    ERR2184202   
site.05.subj.PTAN-0002.lab.TAN-004.iso.1    ERR2184203   
site.05.subj.PTAN-0003.lab.TAN-006.iso.1    ERR2184204   

                                                                                center_name  \
UNIQUEID                                                                                      
site.05.subj.PTAN-0001.lab.TAN-001.iso.1  University of Oxford;Nuffield Department of Me...   
site.05.subj.PTAN-0002.lab.TAN-004.iso.1  University of Oxford;Nuffield Department of Me...   
site.05.subj.PTAN-0003.lab.TAN-006.iso.1  University of Oxford;Nuffield Department of Me...   

                                         country location first_public  \
UNIQUEID                                                                 
site.05.subj.PTAN-0001.lab.TAN-001.iso.1    None     None   2019-06-12   
site.05.subj.PTAN-0002.lab.TAN-004.iso.1    None     None   2019-06-12   
site.05.subj.PTAN-0003.lab.TAN-006.iso.1    None     None   2019-06-12   

                                                                                  fastq_ftp  \
UNIQUEID                                                                                      
site.05.subj.PTAN-0001.lab.TAN-001.iso.1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR218/002/ERR218...   
site.05.subj.PTAN-0002.lab.TAN-004.iso.1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR218/003/ERR218...   
site.05.subj.PTAN-0003.lab.TAN-006.iso.1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR218/004/ERR218...   

                                                                                  fastq_md5  \
UNIQUEID                                                                                      
site.05.subj.PTAN-0001.lab.TAN-001.iso.1  612e42126398723aa1ee84dd1e06feec;349b4db6faead...   
site.05.subj.PTAN-0002.lab.TAN-004.iso.1  469becc8e8fb3961bde32b5352858190;b13aca3a9005c...   
site.05.subj.PTAN-0003.lab.TAN-006.iso.1  d97d6b0ae3ba59f6bfa19e304ac8d94f;022a232fe2db3...   

                                                 fastq_bytes  has_effects  \
UNIQUEID                                                                    
site.05.subj.PTAN-0001.lab.TAN-001.iso.1  79400918;105468452        False   
site.05.subj.PTAN-0002.lab.TAN-004.iso.1   67987132;86536083        False   
site.05.subj.PTAN-0003.lab.TAN-006.iso.1   57530779;72231695        False   

                                          has_variants  has_mutations  \
UNIQUEID                                                                
site.05.subj.PTAN-0001.lab.TAN-001.iso.1         False          False   
site.05.subj.PTAN-0002.lab.TAN-004.iso.1         False          False   
site.05.subj.PTAN-0003.lab.TAN-006.iso.1         False          False   

                                          has_predictions  has_main_report  \
UNIQUEID                                                                     
site.05.subj.PTAN-0001.lab.TAN-001.iso.1            False            False   
site.05.subj.PTAN-0002.lab.TAN-004.iso.1            False            False   
site.05.subj.PTAN-0003.lab.TAN-006.iso.1            False            False   

                                          in_mapping_file  
UNIQUEID                                                   
site.05.subj.PTAN-0001.lab.TAN-001.iso.1            False  
site.05.subj.PTAN-0002.lab.TAN-004.iso.1            False  
site.05.subj.PTAN-0003.lab.TAN-006.iso.1            False

In [7]:
master[master.has_main_report]

study_accession sample_accession  \
UNIQUEID                                                                    
site.04.subj.00042.lab.631456.iso.1           PRJEB41116     SAMEA7524735   
site.ENA.subj.SAMEA104394706.lab.1.iso.1       PRJEB9680   SAMEA104394706   
site.ENA.subj.SAMN08215132.lab.1.iso.1       PRJNA413593     SAMN08215132   
site.ENA.subj.SAMN03647408.lab.1.iso.1       PRJNA282721     SAMN03647408   
site.ENA.subj.SAMN03647375.lab.1.iso.1       PRJNA282721     SAMN03647375   
...                                                  ...              ...   
site.ENA.subj.SAMEA2534732.lab.1.iso.1         PRJEB9680     SAMEA2534732   
site.ENA.subj.SAMEA2534732.lab.1.iso.1         PRJEB9680     SAMEA2534732   
site.ENA.subj.SAMEA2534732.lab.1.iso.1         PRJEB9680     SAMEA2534732   
site.ENA.subj.SAMN07766100.lab.1.iso.1       PRJNA413593     SAMN07766100   
site.ENA.subj.SAMN07766210.lab.1.iso.1       PRJNA413593     SAMN07766210   

                                         run_accession  \
UNIQUEID                                                 
site.04.subj.00042.lab.631456.iso.1         ERR4796279   
site.ENA.subj.SAMEA104394706.lab.1.iso.1    ERR2200134   
site.ENA.subj.SAMN08215132.lab.1.iso.1      SRR6397368   
site.ENA.subj.SAMN03647408.lab.1.iso.1      SRR2100235   
site.ENA.subj.SAMN03647375.lab.1.iso.1      SRR2100202   
...                                                ...   
site.ENA.subj.SAMEA2534732.lab.1.iso.1       ERR552560   
site.ENA.subj.SAMEA2534732.lab.1.iso.1       ERR552559   
site.ENA.subj.SAMEA2534732.lab.1.iso.1       ERR552561   
site.ENA.subj.SAMN07766100.lab.1.iso.1      SRR6152831   
site.ENA.subj.SAMN07766210.lab.1.iso.1      SRR6153023   

                                                                                center_name  \
UNIQUEID                                                                                      
site.04.subj.00042.lab.631456.iso.1       University of Oxford;Nuffield Department of Me...   
site.ENA.subj.SAMEA104394706.lab.1.iso.1                      FORSCHUNGSZENTRUM BORSTEL;FZB   
site.ENA.subj.SAMN08215132.lab.1.iso.1                                           SUB3191708   
site.ENA.subj.SAMN03647408.lab.1.iso.1                                                 None   
site.ENA.subj.SAMN03647375.lab.1.iso.1                                                 None   
...                                                                                     ...   
site.ENA.subj.SAMEA2534732.lab.1.iso.1                        FORSCHUNGSZENTRUM BORSTEL;FZB   
site.ENA.subj.SAMEA2534732.lab.1.iso.1                        FORSCHUNGSZENTRUM BORSTEL;FZB   
site.ENA.subj.SAMEA2534732.lab.1.iso.1                        FORSCHUNGSZENTRUM BORSTEL;FZB   
site.ENA.subj.SAMN07766100.lab.1.iso.1                                                 None   
site.ENA.subj.SAMN07766210.lab.1.iso.1                                                 None   

                                                           country  \
UNIQUEID                                                             
site.04.subj.00042.lab.631456.iso.1                           None   
site.ENA.subj.SAMEA104394706.lab.1.iso.1                      None   
site.ENA.subj.SAMN08215132.lab.1.iso.1    Canada: British Columbia   
site.ENA.subj.SAMN03647408.lab.1.iso.1                     missing   
site.ENA.subj.SAMN03647375.lab.1.iso.1                     missing   
...                                                            ...   
site.ENA.subj.SAMEA2534732.lab.1.iso.1                        None   
site.ENA.subj.SAMEA2534732.lab.1.iso.1                        None   
site.ENA.subj.SAMEA2534732.lab.1.iso.1                        None   
site.ENA.subj.SAMN07766100.lab.1.iso.1    Canada: British Columbia   
site.ENA.subj.SAMN07766210.lab.1.iso.1    Canada: British Columbia   

                                                 location first_public  \
UNIQUEID                                                       

In [9]:
master.index.nunique()

45420

In [53]:
mutations = pandas.read_parquet('mutations.parquet')
# df.reset_index(inplace=True)
mutations[:3]

uniqueid  gene mutation  ref  alt  \
0  site.ENA.subj.SAMN07766182.lab.1.iso.1   PE1    L485L  ctg  ttg   
1  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE14    A106A  gca  gcc   
2  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE16     A96A  gcg  gct   

   nucleotide_number  nucleotide_index  gene_position  codes_protein  \
0                NaN               NaN          485.0           True   
1                NaN               NaN          106.0           True   
2                NaN               NaN           96.0           True   

   indel_length indel_nucleotides  amino_acid_number amino_acid_sequence  \
0           NaN              None              485.0                   L   
1           NaN              None              106.0                   A   
2           NaN              None               96.0                   A   

   number_nucleotide_changes  
0                        1.0  
1                        1.0  
2                        1.0

In [54]:
mutations.info()

<class 'pandas.core.frame.DataFrame'>
Index: 116853 entries, 0 to 2993
Data columns (total 14 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   uniqueid                   116853 non-null  object 
 1   gene                       116853 non-null  object 
 2   mutation                   116853 non-null  object 
 3   ref                        106043 non-null  object 
 4   alt                        106043 non-null  object 
 5   nucleotide_number          9017 non-null    float64
 6   nucleotide_index           9017 non-null    float64
 7   gene_position              116670 non-null  float64
 8   codes_protein              116853 non-null  bool   
 9   indel_length               10598 non-null   float64
 10  indel_nucleotides          10598 non-null   object 
 11  amino_acid_number          99221 non-null   float64
 12  amino_acid_sequence        99221 non-null   object 
 13  number_nucleotide_changes  106043 no

In [55]:
mutations.mutation.value_counts()

mutation
a-3g        155
A111A       146
L63L        144
A72A        125
A18A        122
           ... 
Q553E:65      1
T285T:24      1
I190T:55      1
H255Y:30      1
V145V         1
Name: count, Length: 32541, dtype: int64

In [56]:
mutations[mutations.mutation.str.contains(':')]

uniqueid       gene   mutation  ref  \
9       site.ENA.subj.SAMN07766182.lab.1.iso.1       PE35    E99!:15  gaa   
21      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10   A180A:10  gcc   
22      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10   R225G:22  agg   
23      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    G509G:3  ggc   
24      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    A518G:4  gcc   
...                                        ...        ...        ...  ...   
2511  site.ENA.subj.SAMEA104394448.lab.1.iso.1      mmpL5    T794I:6  acc   
2684  site.ENA.subj.SAMEA104394448.lab.1.iso.1      pks12  A3586V:46  gcg   
2685  site.ENA.subj.SAMEA104394448.lab.1.iso.1      pks12  G3602G:60  ggt   
2720  site.ENA.subj.SAMEA104394448.lab.1.iso.1       plcC     Y97N:3  tac   
2854  site.ENA.subj.SAMEA104394448.lab.1.iso.1       rpfA    A209A:4  gca   

      alt  nucleotide_number  nucleotide_index  gene_position  codes_protein  \
9     taa                NaN               NaN           99.0           True   
21    gcg                NaN               NaN          180.0           True   
22    ggg                NaN               NaN          225.0           True   
23    ggt                NaN               NaN          509.0           True   
24    ggc                NaN               NaN          518.0           True   
...   ...                ...               ...            ...            ...   
2511  atc                NaN               NaN          794.0           True   
2684  gtg                NaN               NaN         3586.0           True   
2685  ggc                NaN               NaN         3602.0           True   
2720  aac                NaN               NaN           97.0           True   
2854  gcg                NaN               NaN          209.0           True   

      indel_length indel_nucleotides  amino_acid_number amino_acid_sequence  \
9              NaN              None               99.0                   !   
21             NaN              None              180.0                   A   
22             NaN              None              225.0                   G   
23             NaN              None              509.0                   G   
24             NaN              None              518.0                   G   
...            ...               ...                ...                 ...   
2511           NaN              None              794.0                   I   
2684           NaN              None             3586.0                   V   
2685           NaN              None             3602.0                   G   
2720           NaN              None               97.0                   N   
2854           NaN              None              209.0                   A   

      number_nucleotide_changes  
9                           1.0  
21                          1.0  
22                          1.0  
23                          1.0  
24                          1.0  
...                         ...  
2511                        1.0  
2684                        1.0  
2685                        1.0  
2720                        1.0  
2854                        1.0  

[17453 rows x 14 columns]

In [69]:
variants = pandas.read_parquet('variants.parquet')
variants[:3]

uniqueid  variant  gene  gene_position  \
0  site.ENA.subj.SAMN01908825.lab.1.iso.1  1977a>g  dnaN          -75.0   
1  site.ENA.subj.SAMN01908825.lab.1.iso.1  4013t>c  recF          245.0   
2  site.ENA.subj.SAMN01908825.lab.1.iso.1  7362g>c  gyrA           21.0   

   codon_idx  nucleotide_index  indel_length indel_nucleotides  \
0        NaN              1977             0              None   
1        1.0              4013             0              None   
2        0.0              7362             0              None   

                                        vcf_evidence  vcf_idx  
0  {"DP": 115, "GT": [1, 1], "GT_CONF": 738.27, "...      1.0  
1  {"ALLELE_DP": [0, 110], "COV": [0, 110], "DP":...      1.0  
2  {"ALLELE_DP": [0, 127], "GT": [1, 1], "DP": 12...      1.0

In [58]:
MUTATIONS = pandas.read_pickle('/Users/fowler/files/cryptic/cryptic-tables-v2.1.2/MUTATIONS.pkl.gz')
MUTATIONS.reset_index(inplace=True)
MUTATIONS[:3]

UNIQUEID     GENE MUTATION  REF  ALT  \
0  site.01.subj.DR0682.lab.DR0682.iso.1   fadD34     S16W  tcg  tgg   
1  site.01.subj.DR0682.lab.DR0682.iso.1   fadD34     T26S  act  tct   
2  site.01.subj.DR0682.lab.DR0682.iso.1  Rv1231c     L60L  ctg  ctc   

   NUCLEOTIDE_NUMBER  NUCLEOTIDE_INDEX  GENE_POSITION  CODES_PROTEIN  \
0                NaN               NaN           16.0           True   
1                NaN               NaN           26.0           True   
2                NaN               NaN           60.0           True   

   INDEL_LENGTH INDEL_NUCLEOTIDES  AMINO_ACID_NUMBER AMINO_ACID_SEQUENCE  \
0           NaN               nan               16.0                   W   
1           NaN               nan               26.0                   S   
2           NaN               nan               60.0                   L   

   NUMBER_NUCLEOTIDE_CHANGES  IS_MINOR_ALLELE MINOR_MUTATION  FRS  
0                        1.0            False           None  NaN  
1                        1.0            False           None  NaN  
2                        1.0            False           None  NaN

In [59]:
VARIANTS = pandas.read_pickle('/Users/fowler/files/cryptic/cryptic-tables-v2.1.2/VARIANTS.pkl.gz')
VARIANTS.reset_index(inplace=True)
VARIANTS[:3]    

UNIQUEID  VARIANT  GENE  GENE_POSITION  \
0  site.01.subj.DR0682.lab.DR0682.iso.1  1977a>g  dnaN          -75.0   
1  site.01.subj.DR0682.lab.DR0682.iso.1  4013t>c  recF          245.0   
2  site.01.subj.DR0682.lab.DR0682.iso.1  7362g>c  gyrA           21.0   

   CODON_IDX  NUCLEOTIDE_INDEX  INDEL_LENGTH INDEL_NUCLEOTIDES  \
0        NaN            1977.0           0.0               nan   
1        1.0            4013.0           0.0               nan   
2        0.0            7362.0           0.0               nan   

                                        VCF_EVIDENCE  VCF_IDX  \
0  {"GT": [1, 1], "DP": 45.0, "ALLELE_DP": [0.0, ...      1.0   
1  {"GT": [1, 1], "DP": 103.0, "ALLELE_DP": [0.0,...      1.0   
2  {"GT": [1, 1], "DP": 61.0, "ALLELE_DP": [0.0, ...      1.0   

   IS_MINOR_ALLELE MINOR_VARIANT  FRS  
0            False          None  NaN  
1            False          None  NaN  
2            False          None  NaN

In [87]:
MUTATIONS.MUTATION.value_counts()[:20]

MUTATION
a-3g       89631
A111A      83541
L63L       80648
A18A       77389
A72A       71442
I145M      65802
A52A       64913
S133S      61820
R21R       60386
A94A       57924
A96A       57782
Q68H       57712
del_1.0    56643
R180G      56503
V210V      55975
A32A       55618
V235V      55091
Q20L       54801
A266A      54601
c-61t      52951
Name: count, dtype: int64

In [75]:
MUTATIONS.columns

Index(['UNIQUEID', 'GENE', 'MUTATION', 'REF', 'ALT', 'NUCLEOTIDE_NUMBER',
       'NUCLEOTIDE_INDEX', 'GENE_POSITION', 'CODES_PROTEIN', 'INDEL_LENGTH',
       'INDEL_NUCLEOTIDES', 'AMINO_ACID_NUMBER', 'AMINO_ACID_SEQUENCE',
       'NUMBER_NUCLEOTIDE_CHANGES', 'IS_MINOR_ALLELE', 'MINOR_MUTATION',
       'FRS'],
      dtype='object')

In [76]:
MUTATIONS[MUTATIONS.IS_MINOR_ALLELE & (MUTATIONS.INDEL_LENGTH>0)]

UNIQUEID    GENE  \
121146                 site.01.subj.DR0623.lab.DR0623.iso.1  Rv0678   
121148                 site.01.subj.DR0623.lab.DR0623.iso.1  Rv0678   
144304                 site.01.subj.DR0144.lab.DR0144.iso.1     gid   
207468                     site.16.subj.001.lab.10398.iso.1    pncA   
210143                     site.16.subj.001.lab.15323.iso.1    inhA   
...                                                     ...     ...   
57561611  site.07.subj.C32A080B-63AC-4427-9257-073281798...     rrs   
57561612  site.07.subj.C32A080B-63AC-4427-9257-073281798...     rrs   
57561616  site.07.subj.C32A080B-63AC-4427-9257-073281798...     rrs   
57561618  site.07.subj.C32A080B-63AC-4427-9257-073281798...     rrs   
57561620  site.07.subj.C32A080B-63AC-4427-9257-073281798...     rrs   

                MUTATION  REF  ALT  NUCLEOTIDE_NUMBER  NUCLEOTIDE_INDEX  \
121146    135_minorindel  nan  nan              135.0          779124.0   
121148    192_minorindel  nan  nan              192.0          779181.0   
144304    352_minorindel  nan  nan              352.0         4407851.0   
207468    457_minorindel  nan  nan              457.0         2288785.0   
210143    -10_minorindel  nan  nan              -10.0         1674192.0   
...                  ...  ...  ...                ...               ...   
57561611  459_minorindel  nan  nan              459.0         1472304.0   
57561612  461_minorindel  nan  nan              461.0         1472306.0   
57561616  483_minorindel  nan  nan              483.0         1472328.0   
57561618  589_minorindel  nan  nan              589.0         1472434.0   
57561620  601_minorindel  nan  nan              601.0         1472446.0   

          GENE_POSITION  CODES_PROTEIN  INDEL_LENGTH INDEL_NUCLEOTIDES  \
121146            135.0           True           5.0             tgtga   
121148            192.0           True           1.0                 g   
144304            352.0           True           1.0                 g   
207468            457.0           True           1.0                 c   
210143            -10.0          False          11.0       taaggatattg   
...                 ...            ...           ...               ...   
57561611          459.0          False           3.0               aca   
57561612          461.0          False           2.0                gg   
57561616          483.0          False           2.0                gt   
57561618          589.0          False           1.0                 c   
57561620          601.0          False           1.0                 c   

          AMINO_ACID_NUMBER AMINO_ACID_SEQUENCE  NUMBER_NUCLEOTIDE_CHANGES  \
121146                  NaN                 nan                        0.0   
121148                  NaN                 nan                        0.0   
144304                  NaN                 nan                        0.0   
207468                  NaN                 nan                        0.0   
210143                  NaN                 nan                        0.0   
...                     ...                 ...                        ...   
57561611                NaN                 nan                        0.0   
57561612                NaN                 nan                        0.0   
57561616                NaN                 nan                        0.0   
57561618                NaN                 nan                        0.0   
57561620                NaN                 nan                        0.0   

          IS_MINOR_ALLELE       MINOR_MUTATION    FRS  
121146               True        135_ins_tgtga  0.310  
121148               True            192_ins_g  0.218  
144304               True            352_ins_g  0.250  
207468               True            457_ins_c  0.851  
210143               True  -10_ins_taaggatattg  0.663  
...                   ...                  ...    ...  
57561611             True          459_ins_aca  0.727  
57561612             Tr

In [90]:
MUTATIONS[MUTATIONS.MUTATION.str[-1]=='x']

Empty DataFrame
Columns: [UNIQUEID, GENE, MUTATION, REF, ALT, NUCLEOTIDE_NUMBER, NUCLEOTIDE_INDEX, GENE_POSITION, CODES_PROTEIN, INDEL_LENGTH, INDEL_NUCLEOTIDES, AMINO_ACID_NUMBER, AMINO_ACID_SEQUENCE, NUMBER_NUCLEOTIDE_CHANGES, IS_MINOR_ALLELE, MINOR_MUTATION, FRS]
Index: []

In [94]:
mutations[mutations.mutation.str.contains('X')][:10]

uniqueid       gene mutation  ref  alt  \
16   site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    A511X  gcc  gxc   
17   site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    A513X  gcg  gxg   
18   site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    L515X  ttg  ttx   
19   site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    A517X  gcc  xcc   
62   site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS23    G443X  ggc  ggx   
112  site.ENA.subj.SAMN07766182.lab.1.iso.1   PE_PGRS4    !838X  taa  tax   
128  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS47    S271X  agc  xgc   
144  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS54    G594X  ggg  ggx   
162  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS56    D177X  gat  gxt   
163  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS57    G265X  ggc  ggx   

     nucleotide_number  nucleotide_index  gene_position  codes_protein  \
16                 NaN               NaN          511.0           True   
17                 NaN               NaN          513.0           True   
18                 NaN               NaN          515.0           True   
19                 NaN               NaN          517.0           True   
62                 NaN               NaN          443.0           True   
112                NaN               NaN          838.0           True   
128                NaN               NaN          271.0           True   
144                NaN               NaN          594.0           True   
162                NaN               NaN          177.0           True   
163                NaN               NaN          265.0           True   

     indel_length indel_nucleotides  amino_acid_number amino_acid_sequence  \
16            NaN              None              511.0                   X   
17            NaN              None              513.0                   X   
18            NaN              None              515.0                   X   
19            NaN              None              517.0                   X   
62            NaN              None              443.0                   X   
112           NaN              None              838.0                   X   
128           NaN              None              271.0                   X   
144           NaN              None              594.0                   X   
162           NaN              None              177.0                   X   
163           NaN              None              265.0                   X   

     number_nucleotide_changes  
16                         1.0  
17                         1.0  
18                         1.0  
19                         1.0  
62                         1.0  
112                        1.0  
128                        1.0  
144                        1.0  
162                        1.0  
163                        1.0

In [83]:
mutations[mutations.mutation.str.contains(':')] # & (mutations.indel_length==0)]

uniqueid       gene   mutation  ref  \
9       site.ENA.subj.SAMN07766182.lab.1.iso.1       PE35    E99!:15  gaa   
21      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10   A180A:10  gcc   
22      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10   R225G:22  agg   
23      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    G509G:3  ggc   
24      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    A518G:4  gcc   
...                                        ...        ...        ...  ...   
2511  site.ENA.subj.SAMEA104394448.lab.1.iso.1      mmpL5    T794I:6  acc   
2684  site.ENA.subj.SAMEA104394448.lab.1.iso.1      pks12  A3586V:46  gcg   
2685  site.ENA.subj.SAMEA104394448.lab.1.iso.1      pks12  G3602G:60  ggt   
2720  site.ENA.subj.SAMEA104394448.lab.1.iso.1       plcC     Y97N:3  tac   
2854  site.ENA.subj.SAMEA104394448.lab.1.iso.1       rpfA    A209A:4  gca   

      alt  nucleotide_number  nucleotide_index  gene_position  codes_protein  \
9     taa                NaN               NaN           99.0           True   
21    gcg                NaN               NaN          180.0           True   
22    ggg                NaN               NaN          225.0           True   
23    ggt                NaN               NaN          509.0           True   
24    ggc                NaN               NaN          518.0           True   
...   ...                ...               ...            ...            ...   
2511  atc                NaN               NaN          794.0           True   
2684  gtg                NaN               NaN         3586.0           True   
2685  ggc                NaN               NaN         3602.0           True   
2720  aac                NaN               NaN           97.0           True   
2854  gcg                NaN               NaN          209.0           True   

      indel_length indel_nucleotides  amino_acid_number amino_acid_sequence  \
9              NaN              None               99.0                   !   
21             NaN              None              180.0                   A   
22             NaN              None              225.0                   G   
23             NaN              None              509.0                   G   
24             NaN              None              518.0                   G   
...            ...               ...                ...                 ...   
2511           NaN              None              794.0                   I   
2684           NaN              None             3586.0                   V   
2685           NaN              None             3602.0                   G   
2720           NaN              None               97.0                   N   
2854           NaN              None              209.0                   A   

      number_nucleotide_changes  
9                           1.0  
21                          1.0  
22                          1.0  
23                          1.0  
24                          1.0  
...                         ...  
2511                        1.0  
2684                        1.0  
2685                        1.0  
2720                        1.0  
2854                        1.0  

[17453 rows x 14 columns]

mutation -> mutation, is_null, is_minor_allele, minor_mutation, minor_reads

`1241_ins_cgccggcgg:10` parsed to `1241_minorindel`, `False`, `True`, `1241_ins_cgccggcgg`, `10`

`E99!:15` parsed to `E99Z`, `False`, `True`, `E99!`, `15`

`g-57x` parsed to `g-57x`, `True`, `False`, None, None

`A511X` parsed to `g-A511X`, `True`, `False`, None, None


In [ ]:
mutation 

def parse_mutations(row):
    if ':' in row.mutation:
        if indel_length > 0 or indel_length < 0:
            mut
        mut,minor_reads = row.mutation.split(':')
        is_minor = True
    else:
        mut = row.mutation
        minor_reads = None
        is_minor = False

    return [row.mutation]

In [72]:
VARIANTS[VARIANTS.IS_MINOR_ALLELE][:10]

UNIQUEID     VARIANT  GENE  GENE_POSITION  \
7950   site.01.subj.DR0611.lab.DR0611.iso.1   760729a>z  rpoB          308.0   
16139  site.01.subj.DR0600.lab.DR0600.iso.1  1472805c>z   rrs          960.0   
34546  site.01.subj.DR0685.lab.DR0685.iso.1     7572t>z  gyrA           91.0   
34547  site.01.subj.DR0685.lab.DR0685.iso.1     7582a>z  gyrA           94.0   
41351  site.01.subj.DR0653.lab.DR0653.iso.1     8164c>z  gyrA          288.0   
79166  site.01.subj.DR0589.lab.DR0589.iso.1   761100c>z  rpoB          432.0   
79167  site.01.subj.DR0589.lab.DR0589.iso.1   761139c>z  rpoB          445.0   
79168  site.01.subj.DR0589.lab.DR0589.iso.1   761161t>z  rpoB          452.0   
93493  site.01.subj.DR0679.lab.DR0679.iso.1     7581g>z  gyrA           94.0   
93494  site.01.subj.DR0679.lab.DR0679.iso.1     7582a>z  gyrA           94.0   

       CODON_IDX  NUCLEOTIDE_INDEX  INDEL_LENGTH INDEL_NUCLEOTIDES  \
7950         2.0          760729.0           0.0               nan   
16139        NaN         1472805.0           0.0               nan   
34546        1.0            7572.0           0.0               nan   
34547        2.0            7582.0           0.0               nan   
41351        2.0            8164.0           0.0               nan   
79166        1.0          761100.0           0.0               nan   
79167        1.0          761139.0           0.0               nan   
79168        2.0          761161.0           0.0               nan   
93493        1.0            7581.0           0.0               nan   
93494        2.0            7582.0           0.0               nan   

                                            VCF_EVIDENCE  VCF_IDX  \
7950   {"GT": [1, 1], "DP": 77.0, "ALLELE_DP": [12.0,...      1.0   
16139  {"GT": [0, 0], "DP": 47.0, "ALLELE_DP": [47.0,...      1.0   
34546  {"GT": [0, 0], "DP": 36.0, "ALLELE_DP": [36.0,...      1.0   
34547  {"GT": [1, 1], "DP": 39.0, "ALLELE_DP": [24.0,...      1.0   
41351  {"GT": [1, 1], "DP": 121.0, "ALLELE_DP": [17.0...      1.0   
79166  {"GT": [0, 0], "DP": 37.0, "ALLELE_DP": [37.0,...      1.0   
79167  {"GT": [0, 0], "DP": 33.0, "ALLELE_DP": [33.0,...      1.0   
79168  {"GT": [1, 1], "DP": 32.0, "ALLELE_DP": [18.0,...      1.0   
93493  {"GT": [1, 1], "DP": 43.0, "ALLELE_DP": [25.0,...      1.0   
93494  {"GT": [0, 0], "DP": 46.0, "ALLELE_DP": [46.0,...      1.0   

       IS_MINOR_ALLELE MINOR_VARIANT    FRS  
7950              True     760729a>g  0.865  
16139             True    1472805c>t  0.230  
34546             True       7572t>c  0.357  
34547             True       7582a>g  0.619  
41351             True       8164c>a  0.877  
79166             True     761100c>a  0.098  
79167             True     761139c>a  0.283  
79168             True     761161t>c  0.640  
93493             True       7581g>a  0.632  
93494             True       7582a>g  0.352

In [105]:
variants[variants.variant.str.contains('x')]

uniqueid     variant       gene  \
45   site.ENA.subj.SAMN01908825.lab.1.iso.1    79571c>x     Rv0071   
46   site.ENA.subj.SAMN01908825.lab.1.iso.1    80616c>x     Rv0072   
144  site.ENA.subj.SAMN01908825.lab.1.iso.1   336560t>x   PE_PGRS4   
145  site.ENA.subj.SAMN01908825.lab.1.iso.1   336590g>x   PE_PGRS4   
146  site.ENA.subj.SAMN01908825.lab.1.iso.1   336592t>x   PE_PGRS4   
..                                      ...         ...        ...   
721  site.ENA.subj.SAMN07659930.lab.1.iso.1  3739498c>x  PE_PGRS50   
762  site.ENA.subj.SAMN07659930.lab.1.iso.1  3934022t>x  PE_PGRS54   
764  site.ENA.subj.SAMN07659930.lab.1.iso.1  3935950t>x  PE_PGRS54   
765  site.ENA.subj.SAMN07659930.lab.1.iso.1  3946813a>x  PE_PGRS57   
766  site.ENA.subj.SAMN07659930.lab.1.iso.1  3948612a>x  PE_PGRS57   

     gene_position  codon_idx  nucleotide_index  indel_length  \
45            29.0        1.0             79571             0   
46            -8.0        NaN             80616             0   
144          838.0        2.0            336560             0   
145          828.0        2.0            336590             0   
146          828.0        0.0            336592             0   
..             ...        ...               ...           ...   
721         1093.0        0.0           3739498             0   
762         1006.0        2.0           3934022             0   
764         1649.0        1.0           3935950             0   
765          340.0        2.0           3946813             0   
766          940.0        1.0           3948612             0   

    indel_nucleotides                                       vcf_evidence  \
45               None  {"ALLELE_DP": [0, 0], "FRS": ".", "GT_CONF_PER...   
46               None  {"GT": [1, 1], "FRS": 1.0, "ALLELE_DP": [0, 1]...   
144              None  {"GT": [0, 0], "COV_TOTAL": 1, "GT_CONF": 7.69...   
145              None  {"FRS": 0.5, "GT": [0, 0], "GT_CONF": 0.0, "CO...   
146              None  {"FRS": 0.5, "COV": [1, 1], "COV_TOTAL": 2, "G...   
..                ...                                                ...   
721              None  {"GT_CONF": 22.3, "ALLELE_DP": [0, 4], "DP": 4...   
762              None  {"GT": [1, 1], "GT_CONF_PERCENTILE": 0.04, "CO...   
764              None  {"COV_TOTAL": 0, "COV": [0, 0], "GT_CONF": 0.0...   
765              None  {"GT_CONF_PERCENTILE": 0.04, "COV": [0, 1], "A...   
766              None  {"COV_TOTAL": 3, "GT_CONF": 16.0, "DP": 2, "GT...   

     vcf_idx  
45       NaN  
46       1.0  
144      0.0  
145      0.0  
146      0.0  
..       ...  
721      1.0  
762      1.0  
764      NaN  
765      1.0  
766      1.0  

[6083 rows x 10 columns]

variant -> variant, is_null, is_minor_allele, minor_varian, minor_reads, total_reads

`7570c>t:76` parsed to `7570c>z`, `False`, `True`, `7570c>t`, 76, 110

`24699_del_ccgcgttgctcggggtaa:41` parsed to `24699_minorindel`, `False`, `True`, `24699_del_ccgcgttgctcggggtaa`, 41, 47

`338554t>x` parsed to `338554t>x`, `True`, `False`, None, None			


```
"{""COV"": [34, 76], ""COV_TOTAL"": 110, ""GT"": [1, 1], ""GT_CONF"": 261.86, ""GT_CONF_PERCENTILE"": 6.47, ""ALLELE_DP"": [34, 76], ""DP"": 76, ""FRS"": 0.6909, ""POS"": 7570, ""REF"": ""c"", ""ALTS"": [""t""]}"
```

In [121]:
import json 

def parse_variants(row):

    variant = row.variant
    is_null = False
    is_minor = False
    minor_variant = None
    minor_reads = None
    coverage = None
    frs = None

    a=(json.loads(row['vcf_evidence']))
    if 'COV_TOTAL' in a.keys():
        coverage = a['COV_TOTAL']
    if 'FRS' in a.keys():
        frs = a['FRS']

    if row.variant[-1]=='x':
        is_null = True
    elif ':' in row.variant:
        is_minor = True
        minor_variant, minor_reads = row.variant.split(':')
        if "ins" in variant or "del" in variant:
            variant = minor_variant.split('_')[0]+"_minorindel"
        else:
            variant = minor_variant[:-1]+'x'

    return pandas.Series([variant, is_null, is_minor, minor_variant, minor_reads, coverage, frs])

variants[['var','is_null','is_minor','minor_variant','minor_reads','coverage','FRS']] = variants.apply(parse_variants, axis=1)

In [122]:
variants

uniqueid                  variant  \
0    site.ENA.subj.SAMN01908825.lab.1.iso.1                  1977a>g   
1    site.ENA.subj.SAMN01908825.lab.1.iso.1                  4013t>c   
2    site.ENA.subj.SAMN01908825.lab.1.iso.1                  7362g>c   
3    site.ENA.subj.SAMN01908825.lab.1.iso.1                  7585g>c   
4    site.ENA.subj.SAMN01908825.lab.1.iso.1                  9304g>a   
..                                      ...                      ...   
916  site.ENA.subj.SAMN07659930.lab.1.iso.1             3948649a>c:4   
917  site.ENA.subj.SAMN07659930.lab.1.iso.1             4032006c>g:4   
918  site.ENA.subj.SAMN07659930.lab.1.iso.1  4036926_del_gccgcccgc:4   
919  site.ENA.subj.SAMN07659930.lab.1.iso.1            4060210t>a:61   
920  site.ENA.subj.SAMN07659930.lab.1.iso.1             4263501t>g:9   

          gene  gene_position  codon_idx  nucleotide_index  indel_length  \
0         dnaN          -75.0        NaN              1977             0   
1         recF          245.0        1.0              4013             0   
2         gyrA           21.0        0.0              7362             0   
3         gyrA           95.0        1.0              7585             0   
4         gyrA          668.0        1.0              9304             0   
..         ...            ...        ...               ...           ...   
916  PE_PGRS57          952.0        2.0           3948649             0   
917  PE_PGRS58          385.0        0.0           4032006             0   
918  PE_PGRS59         1117.0        2.0           4036926            -9   
919       esxV           20.0        1.0           4060210             0   
920    Rv3802c          289.0        0.0           4263501             0   

    indel_nucleotides                                       vcf_evidence  \
0                None  {"DP": 115, "GT": [1, 1], "GT_CONF": 738.27, "...   
1                None  {"ALLELE_DP": [0, 110], "COV": [0, 110], "DP":...   
2                None  {"ALLELE_DP": [0, 127], "GT": [1, 1], "DP": 12...   
3                None  {"ALLELE_DP": [0, 114], "COV_TOTAL": 114, "GT"...   
4                None  {"GT_CONF_PERCENTILE": 68.63, "COV_TOTAL": 114...   
..                ...                                                ...   
916              None  {"COV_TOTAL": 7, "FRS": 0.5714, "GT": [1, 1], ...   
917              None  {"COV": [0, 4], "GT": [1, 1], "DP": 4, "ALLELE...   
918         gccgcccgc  {"GT_CONF": 36.62, "COV_TOTAL": 4, "GT_CONF_PE...   
919              None  {"COV": [9, 61], "FRS": 0.8714, "GT_CONF_PERCE...   
920              None  {"GT_CONF": 311.55, "GT_CONF_PERCENTILE": 52.8...   

     vcf_idx                 var  is_null  is_minor          minor_variant  \
0        1.0             1977a>g    False     False                   None   
1        1.0             4013t>c    False     False                   None   
2        1.0             7362g>c    False     False                   None   
3        1.0             7585g>c    False     False                   None   
4        1.0             9304g>a    False     False                   None   
..       ...                 ...      ...       ...                    ...   
916      1.0          3948649a>x    False      True             3948649a>c   
917      1.0          4032006c>x    False      True             4032006c>g   
918      1.0  4036926_minorindel    False      True  4036926_del_gccgcccgc   
919      1.0          4060210t>x    False      True             4060210t>a   
920      1.0          4263501t>x    False      True             4263501t>g   

    minor_reads  coverage     FRS  
0          None     115.0     1.0  
1          None     110.0     1.0  
2          None     127.0     1.0  
3          None     114.0     1.0  
4          None     114.0     1.0  
..          ...       ...     ...  
916           4       7.0  0.5714  
917           4       4.0     1.0  
918           4       4.0     1.0  
919          61      70.0  0.8714  
920        

In [124]:

variants[variants.is_minor]

uniqueid                          variant  \
1542  site.ENA.subj.SAMN01908825.lab.1.iso.1                       7570c>t:76   
1543  site.ENA.subj.SAMN01908825.lab.1.iso.1                       7572t>c:35   
1544  site.ENA.subj.SAMN01908825.lab.1.iso.1  24699_del_ccgcgttgctcggggtaa:41   
1545  site.ENA.subj.SAMN01908825.lab.1.iso.1                       24721g>c:7   
1546  site.ENA.subj.SAMN01908825.lab.1.iso.1                       24722t>c:7   
...                                      ...                              ...   
916   site.ENA.subj.SAMN07659930.lab.1.iso.1                     3948649a>c:4   
917   site.ENA.subj.SAMN07659930.lab.1.iso.1                     4032006c>g:4   
918   site.ENA.subj.SAMN07659930.lab.1.iso.1          4036926_del_gccgcccgc:4   
919   site.ENA.subj.SAMN07659930.lab.1.iso.1                    4060210t>a:61   
920   site.ENA.subj.SAMN07659930.lab.1.iso.1                     4263501t>g:9   

           gene  gene_position  codon_idx  nucleotide_index  indel_length  \
1542       gyrA           90.0        1.0              7570             0   
1543       gyrA           91.0        0.0              7572             0   
1544       fhaA          729.0        1.0             24699           -18   
1545       fhaA          242.0        0.0             24721             0   
1546       fhaA          241.0        2.0             24722             0   
...         ...            ...        ...               ...           ...   
916   PE_PGRS57          952.0        2.0           3948649             0   
917   PE_PGRS58          385.0        0.0           4032006             0   
918   PE_PGRS59         1117.0        2.0           4036926            -9   
919        esxV           20.0        1.0           4060210             0   
920     Rv3802c          289.0        0.0           4263501             0   

       indel_nucleotides                                       vcf_evidence  \
1542                None  {"COV": [34, 76], "COV_TOTAL": 110, "GT": [1, ...   
1543                None  {"COV": [77, 35], "COV_TOTAL": 112, "GT_CONF":...   
1544  ccgcgttgctcggggtaa  {"COV_TOTAL": 103, "COV": [10, 41], "GT": [1, ...   
1545                None  {"GT": [0, 0], "COV_TOTAL": 85, "FRS": 0.9176,...   
1546                None  {"FRS": 0.9195, "COV": [73, 7], "ALLELE_DP": [...   
...                  ...                                                ...   
916                 None  {"COV_TOTAL": 7, "FRS": 0.5714, "GT": [1, 1], ...   
917                 None  {"COV": [0, 4], "GT": [1, 1], "DP": 4, "ALLELE...   
918            gccgcccgc  {"GT_CONF": 36.62, "COV_TOTAL": 4, "GT_CONF_PE...   
919                 None  {"COV": [9, 61], "FRS": 0.8714, "GT_CONF_PERCE...   
920                 None  {"GT_CONF": 311.55, "GT_CONF_PERCENTILE": 52.8...   

      vcf_idx                 var  is_null  is_minor  \
1542      1.0             7570c>x    False      True   
1543      1.0             7572t>x    False      True   
1544      1.0    24699_minorindel    False      True   
1545      1.0            24721g>x    False      True   
1546      1.0            24722t>x    False      True   
...       ...                 ...      ...       ...   
916       1.0          3948649a>x    False      True   
917       1.0          4032006c>x    False      True   
918       1.0  4036926_minorindel    False      True   
919       1.0          4060210t>x    False      True   
920       1.0          4263501t>x    False      True   

                     minor_variant minor_reads  coverage     FRS  
1542                       7570c>t          76     110.0  0.6909  
1543                       7572t>c          35     112.0  0.6875  
1544  24699_del_ccgcgttgctcggggtaa          41     103.0  0.7767  
1545                      24721g>c           7      85.0  0.9176  
1546                      24722t>c           7      87.0  0.9195  
...                            ...         ...       ...     ...  
916                     3948649a>c           4 